data-upgrade-users/src/test/java/org/exoplatform/migration/UserPasswordHashMigrationTest.java

```diff
@@ -0,0 +1,119 @@
+
+  @Before
+  public void setUp() {
+    container = PortalContainer.getInstance();
+    settingService = container.getComponentInstanceOfType(SettingService.class);
+    secureRandomService = container.getComponentInstanceOfType(SecureRandomService.class);
+    userPasswordHashMigration = new UserPasswordHashMigration(entityManagerService,
+                                                              picketLinkIDMService,
+                                                              secureRandomService,
+                                                              settingService,
+                                                              new InitParams());
+  }
+
+  @Test
+  public void processUpgrade() throws Exception {
+    List<Object[]> result = new ArrayList<>();
+    result.add(new String[] { "user", "passwordHash" });
+    EntityManager entityManager = mock(EntityManager.class);
+    Query query = mock(Query.class);
+    when(entityManager.createNativeQuery(anyString())).thenReturn(query);
+    when(query.getResultList()).thenReturn(result);
+    when(entityManagerService.getEntityManager()).thenReturn(entityManager);
+
+    IdentitySession identitySession = mock(IdentitySession.class);
+    PersistenceManager persistenceManager = mock(PersistenceManager.class);
+    ExtendedAttributeManager extendedAttributeManager = mock(ExtendedAttributeManager.class);
+    when(identitySession.getPersistenceManager()).thenReturn(persistenceManager);
+    when(picketLinkIDMService.getIdentitySession()).thenReturn(identitySession);
+    when(picketLinkIDMService.getExtendedAttributeManager()).thenReturn(extendedAttributeManager);
+    boolean proceedToUpgrade = userPasswordHashMigration.shouldProceedToUpgrade(null, null, null);
+    assertTrue(proceedToUpgrade);
+    SettingValue<?> settingValue = settingService.get(Context.GLOBAL.id(USER_PASSWORD_HASH_MIGRATION),
+                                                      Scope.APPLICATION.id(USER_PASSWORD_HASH_MIGRATION),
+                                                      USER_PASSWORD_HASH_MIGRATION_SETTING_KEY);
+    assertNull(settingValue);
+    userPasswordHashMigration.processUpgrade(null, null);
+    userPasswordHashMigration.afterUpgrade();
+    verify(extendedAttributeManager, times(1)).updatePassword(any(), anyString());
+    verify(extendedAttributeManager, times(1)).addAttribute(anyString(), anyString(), anyString());
+
+    settingValue = settingService.get(Context.GLOBAL.id(USER_PASSWORD_HASH_MIGRATION),
+                                      Scope.APPLICATION.id(USER_PASSWORD_HASH_MIGRATION),
+                                      USER_PASSWORD_HASH_MIGRATION_SETTING_KEY);
+    assertNotNull(settingValue);
+    assertEquals(true, settingValue.getValue());
+
+    proceedToUpgrade = userPasswordHashMigration.shouldProceedToUpgrade(null, null, null);
+    assertFalse(proceedToUpgrade);
+  }
+}
```

Comment:
can you add a test with an error during process upgrade, to check that the upgrade correctly restart ?

Reply:
already exists by not calling the afterUpgrade, the shouldProceedToUpgrade is true, so when inttrepring server start or stop the server during the UP execution , the afterUpgrade will not be called, so the next restart the UP will be launched again .
https://github.com/exoplatform/data-upgrade/pull/161/files#diff-1a9349cb98114358fe6ed02277b76ad326d706337cec3efe4e732b4f6ed39b41R100

Reply:
In fact, my test would be 

boolean upgrade = processUpgrade
//with mock find a way to fail it
assertFalse(upgrade)

//restart processUpgrade
assert(process upgrade correctly restart and finish)